js/src/vector/row.ts

```diff
@@ -15,106 +15,281 @@
+    }
+    public getKey(idx: number): K['TValue'] {
+        return this[kParent].getChildAt(0)!.get(idx);
+    }
+    public getIndex(key: K['TValue']): number {
+        return this[kParent].getChildAt(0)!.indexOf(key);
     }
+    public getValue(index: number): V['TValue'] | null {
+        return this[kParent].getChildAt(1)!.get(index);
+    }
+    public setValue(index: number, value: V['TValue'] | null): void {
+        this[kParent].getChildAt(1)!.set(index, value);
+    }
+}
 
-    private _bindGetter(columnIndex: number) {
-        return function(this: Row<T>) {
-            const child = this[kParent].getChildAt(columnIndex);
-            return child ? child.get(this[kRowIndex]) : null;
-        };
+export class StructRow<T extends { [key: string]: DataType } = any> extends Row<keyof T, T[keyof T]['TValue'] | null> {
+    constructor(parent: StructVector<T>) {
+        super(parent, parent.type.children.length);
+        return defineRowProxyProperties(this);
+    }
+    public *keys() {
+        for (const field of this[kParent].type.children) {
+            yield field.name as keyof T;
+        }
+    }
+    public *values() {
+        for (const field of this[kParent].type.children) {
+            yield (this as RowLike<T>)[field.name];
+        }
+    }
+    public getKey(idx: number): keyof T {
+        return this[kParent].type.children[idx].name as keyof T;
     }
-    public bind(rowIndex: number) {
-        const bound: Row<T> = Object.create(this.rowPrototype);
-        bound[kRowIndex] = rowIndex;
-        return bound;
+    public getIndex(key: keyof T): number {
+        return this[kParent].type.children.findIndex((f) => f.name === key);
+    }
+    public getValue(index: number): T[keyof T]['TValue'] | null {
+        return this[kParent].getChildAt(index)!.get(this[kRowIndex]);
+    }
+    public setValue(index: number, value: T[keyof T]['TValue'] | null): void {
+        return this[kParent].getChildAt(index)!.set(this[kRowIndex], value);
     }
 }
+
+Object.setPrototypeOf(Row.prototype, Map.prototype);
+
+/** @ignore */
+const defineRowProxyProperties = (() => {
+    const desc = { enumerable: true, configurable: false, get: null as any, set: null as any };
+    return <T extends Row>(row: T) => {
+        let idx = -1, ktoi = row[kKeyToIdx] || (row[kKeyToIdx] = new Map());
+        const getter = (key: any) => function(this: T) { return this.get(key); };
+        const setter = (key: any) => function(this: T, val: any) { return this.set(key, val); };
+        for (const key of row.keys()) {
+            ktoi.set(key, ++idx);
+            desc.get = getter(key);
+            desc.set = setter(key);
+            row.hasOwnProperty(key) || (desc.enumerable = true, Object.defineProperty(row, key, desc));
+            row.hasOwnProperty(idx) || (desc.enumerable = false, Object.defineProperty(row, idx, desc));
+        }
+        desc.get = desc.set = null;
+        return row;
+    };
+})();
+
```

Comment:
Could you run the performance tests to make sure this change doesn't affect perf for the row proxies? I don't think it would, but doesn't hurt to make sure.